### ISSUE TITLE
Editor mode for working with a fixed external command

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#602]
Sometimes we need to use one command to do a series of things, notably VCS like `git` and `hg`. Using the example of `git`, it would be useful we if can enter a git mode, and `add` automatically means `git add`, `push` means `git push`, etc., and completions also treat `add` as `git add`.

In fact, I remember there was a project that provided a git CLI with completion which made to the homepage of HN, but I cannot find it anymore...

**Comment by zzamboni** [comment]
This would be great. Please make it usable also for namespaces, i.e. so you could enter "epm mode" and then `install` would automatically mean `epm:install`, etc.

**Comment by xiaq** [comment]
@zzamboni Good point. FWIW, namespaces are now first-class values in Elvish, and they are indexable. Their names are simply the name of the namespace, including the trailing `:`. You can do this:

```elvish-transcript
~> use str
~> put $str:
▶ <ns 0xc42016e9f0>
~> put $str:[compare~]
▶ <builtin str:compare>
```

It is not too hard to write a `str` function that delegates calls to functions in the `str:` namespace:

```elvish-transcript
~> fn str [f @a]{ $str:[$f'~'] $@a }
~> str compare a b
▶ -1
```

**Comment by zzamboni** [comment]
@xiaq this is great news! I wasn't aware this was implemented already. I tried your example and it works! I notice that you can also use this to access variables, but you cannot set them:

```
[~]─> use epm
[~]─> put $epm:-data-dir
▶ /Users/taazadi1/.elvish
[~]─> put $epm:[-data-dir]
▶ /Users/taazadi1/.elvish
[~]─> epm:[-data-dir] = '/foo/bar'
Exception: assoc is not supported
[tty], line 1: epm:[-data-dir] = '/foo/bar'
```

Are you planning to make it possible to modify the namespace as well?

Also, is there a way to list the contents of a namespace? 

```
[~]─> keys $str:
Exception: wrong type of 1'th argument: need !!<nil>, got ns
[~]─> has-key $str: compare
Exception: couldn't get key or index of type 'ns'
```
